File: templates/python/flask/app/api/health.py
"""
Health check endpoints
"""
from flask import Blueprint, jsonify, current_app
from sqlalchemy import text

from app import db, cache
from app.config import Config

health_bp = Blueprint('health', __name__)


@health_bp.route('/health')
def health_check():
    """Basic health check"""
    return jsonify({
        'status': 'healthy',
        'service': Config.SERVICE_NAME,
        'version': '1.0.0'
    })


@health_bp.route('/health/detailed')
def detailed_health_check():
    """Detailed health check with dependencies"""
    health_status = {
        'status': 'healthy',
        'service': Config.SERVICE_NAME,
        'checks': {
            'api': 'healthy',
            'database': 'unknown',
            'redis': 'unknown'
        }
    }
    
    # Check database
    try:
        db.session.execute(text('SELECT 1'))
        health_status['checks']['database'] = 'healthy'
    except Exception as e:
        health_status['checks']['database'] = 'unhealthy'
        health_status['status'] = 'degraded'
        current_app.logger.error(f"Database health check failed: {e}")
    
    # Check Redis/Cache
    try:
        cache.get('health_check')
        cache.set('health_check', 'ok', timeout=10)
        health_status['checks']['redis'] = 'healthy'
    except Exception as e:
        health_status['checks']['redis'] = 'unhealthy'
        health_status['status'] = 'degraded'
        current_app.logger.error(f"Redis health check failed: {e}")
    
    return jsonify(health_status)


@health_bp.route('/ready')
def readiness_check():
    """Readiness check"""
    # Add any startup checks here
    return jsonify({
        'status': 'ready',
        'service': Config.SERVICE_NAME
    })

File: templates/rest-api/src/routes/health.ts
import { Router, Request, Response } from 'express';
import { StatusCodes } from 'http-status-codes';
import os from 'os';
import { config } from '../config';

export const healthRouter = Router();

interface HealthCheck {
  status: 'healthy' | 'unhealthy';
  timestamp: string;
  uptime: number;
  service: string;
  version: string;
  environment: string;
  checks?: Record<string, HealthCheckResult>;
}

interface HealthCheckResult {
  status: 'healthy' | 'unhealthy';
  message?: string;
  latency?: number;
}

// Liveness probe - basic health check
healthRouter.get('/live', (req: Request, res: Response) => {
  res.status(StatusCodes.OK).json({
    status: 'healthy',
    timestamp: new Date().toISOString(),
  });
});

// Readiness probe - detailed health check
healthRouter.get('/ready', async (req: Request, res: Response) => {
  const startTime = Date.now();
  const checks: Record<string, HealthCheckResult> = {};
  
  // Check database connection (example)
  if (config.database.url) {
    try {
      // TODO: Implement actual database ping
      checks.database = {
        status: 'healthy',
        latency: Date.now() - startTime,
      };
    } catch (error) {
      checks.database = {
        status: 'unhealthy',
        message: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }
  
  // Check Redis connection (example)
  if (config.redis.url) {
    try {
      // TODO: Implement actual Redis ping
      checks.redis = {
        status: 'healthy',
        latency: Date.now() - startTime,
      };
    } catch (error) {
      checks.redis = {
        status: 'unhealthy',
        message: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }
  
  // Check memory usage
  const memoryUsage = process.memoryUsage();
  const totalMemory = os.totalmem();
  const memoryPercentage = (memoryUsage.rss / totalMemory) * 100;
  
  checks.memory = {
    status: memoryPercentage < 90 ? 'healthy' : 'unhealthy',
    message: `${memoryPercentage.toFixed(2)}% memory used`,
  };
  
  // Overall health status
  const isHealthy = Object.values(checks).every(check => check.status === 'healthy');
  
  const health: HealthCheck = {
    status: isHealthy ? 'healthy' : 'unhealthy',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    service: 'rest-api',
    version: process.env.npm_package_version || '1.0.0',
    environment: config.env,
    checks,
  };
  
  res.status(isHealthy ? StatusCodes.OK : StatusCodes.SERVICE_UNAVAILABLE).json(health);
});

// Detailed health information
healthRouter.get('/', async (req: Request, res: Response) => {
  const health = {
    status: 'healthy',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    service: 'rest-api',
    version: process.env.npm_package_version || '1.0.0',
    environment: config.env,
    system: {
      platform: os.platform(),
      arch: os.arch(),
      nodeVersion: process.version,
      memory: {
        total: os.totalmem(),
        free: os.freemem(),
        used: process.memoryUsage(),
      },
      cpu: {
        cores: os.cpus().length,
        model: os.cpus()[0]?.model,
        loadAverage: os.loadavg(),
      },
    },
  };
  
  res.status(StatusCodes.OK).json(health);
});